modules/xmpp/JingleSession.js

```diff
@@ -1,148 +1,161 @@
      */
-    this.me = me;
+    constructor(me, sid, peerjid, connection, media_constraints, ice_config) {
+        this.me = me;
+        this.sid = sid;
+        this.peerjid = peerjid;
+        this.connection = connection;
+        this.media_constraints = media_constraints;
+        this.ice_config = ice_config;
+
+        /**
+         * Whether to use dripping or not. Dripping is sending trickle
+         * candidates not one-by-one.
+         * Note: currently we do not support 'false'.
+         */
+        this.usedrip = true;
+
+        /**
+         *  When dripping is used, stores ICE candidates which are to be sent.
+         */
+        this.drip_container = [];
+
+        /**
+         * The chat room instance associated with the session.
+         * @type {ChatRoom}
+         */
+        this.room = null;
+
+        /**
+         * Jingle session state - uninitialized until {@link initialize} is
+         * called @type {JingleSessionState}
+         */
+        this.state = null;
+
+        /**
+         * The RTC service instance
+         * @type {RTC}
+         */
+        this.rtc = null;
+    }
 
     /**
-     * The Jingle session identifier.
+     * Prepares this object to initiate a session.
+     * @param {boolean} isInitiator whether we will be the Jingle initiator.
+     * @param {ChatRoom} room the chat room for the conference associated with
+     * this session
+     * @param {RTC} rtc the RTC service instance
      */
-    this.sid = sid;
+    initialize(isInitiator, room, rtc) {
+        if (this.state !== null) {
+            const errmsg
+                = 'attempt to initiate on session ' + this.sid + 'in state '
+                    + this.state;
+            logger.error(errmsg);
+            throw new Error(errmsg);
+        }
+        this.room = room;
+        this.rtc = rtc;
+        this.state = JingleSessionState.PENDING;
+        this.initiator = isInitiator ? this.me : this.peerjid;
+        this.responder = !isInitiator ? this.me : this.peerjid;
+        this.doInitialize();
+    }
 
     /**
-     * the JID of the remote peer.
+     * The implementing class finishes initialization here. Called at the end of
+     * {@link initialize}.
      */
-    this.peerjid = peerjid;
+    doInitialize() {
+
+    }
 
     /**
-     * The XMPP connection.
+     * Adds the ICE candidates found in the 'contents' array as remote
+     * candidates?
+     * Note: currently only used on transport-info
      */
-    this.connection = connection;
+    // eslint-disable-next-line no-unused-vars
+    addIceCandidates (contents) {
+
+    }
 
     /**
-     * Whether to use dripping or not. Dripping is sending trickle candidates
-     * not one-by-one.
-     * Note: currently we do not support 'false'.
+     * Returns current state of this <tt>JingleSession</tt> instance.
+     * @returns {JingleSessionState} the current state of this session instance.
      */
-    this.usedrip = true;
+    getState () {
```

Comment:
I assume we have this because `this.state` is supposed to be private? If so, a more idiomatic way to do it isn using ES6 getters:

```
get state() {
    return this._state;
}
```

And we make state a private attribute of the instance. Consumers can access it as `session.state` just fine, but they cannot set it.